File: ros/src/util/packages/data_preprocessor/scripts/get_PCD.py
#!/usr/bin/env python
import sys
import os
import rospy
import numpy as np
import cv2
import pcl
import sensor_msgs.point_cloud2 as pc2
from sensor_msgs.msg import PointCloud2
from cv_bridge import CvBridge

save_path = None

def cloud_loader(msg):
    timestamp = msg.header.stamp.secs + ((msg.header.stamp.nsecs + 0.0) / 1000000000)
    save_pcd(msg, timestamp, save_path)

def save_pcd(cloud, timestamp, path):
    p = pcl.PointCloud(np.array(list(pc2.read_points(cloud)), dtype=np.float32)[:, 0:3])
    p.to_file(path + '/pcd' + '_' + "{:.5f}".format(timestamp) + '.pcd')

def rosbag_data_extract_sample():
    global save_path
    try:
        save_path = sys.argv[1]
        topic = sys.argv[2]
    except Exception, e:
        #sys.exit("Please specify the save path. Example: rosbag_data_extract_unsync.py /media/0/output/")
        save_path = './sample'

    node_name = "get_%s_and_convert_to_PCD_data" % topic
    rospy.init_node('rosbag_pcd_extract_unsync', anonymous=True)

    rospy.Subscriber(topic, PointCloud2, cloud_loader)
    rospy.spin()

if __name__ == '__main__':
    rosbag_data_extract_sample()

File: ros/src/util/packages/data_preprocessor/scripts/get_PCD.cpp
#include "get_PCD.hpp"
// catkin_create_pkg pcd_tutorial std_msgs roscpp pcl_ros pcl_msgs pcl_conversions libpcl-all-dev

static void check_arguments(int argc, char* argv[])
{
  if (argc != 3){
    cout << "Please set arguments like below\n'rosrun data_preprocessor get_PCD save_dir topic_name'\n";
    exit(EXIT_FAILURE);
  }
}

void SavePCD::save_pcd(const sensor_msgs::PointCloud2::ConstPtr& msg)
{
  stringstream ss;
  ss << msg->header.stamp;
  string file_name = save_path_ + ss.str() + ".pcd";
  pcl::PointCloud<pcl::PointXYZ> points;
  pcl::fromROSMsg(*msg, points);
  pcl::io::savePCDFileASCII(file_name, points);
}

void SavePCD::sub_pcd(int argc, char* argv[])
{
  ros::init(argc, argv, "PCD_Subscriber");
  ros::NodeHandle n;
  ros::Subscriber sub = n.subscribe(topic_name_, 1, &SavePCD::save_pcd, this);
  ros::spin();
}

int main(int argc, char* argv[])
{
  check_arguments(argc, argv);
  SavePCD si;
  string path = argv[1];
  if (path[path.size() - 1] == '/'){
    path.erase(path.begin() + path.size()-1);
  }
  si.save_path_ = path + '/';
  si.topic_name_ = argv[2];
  si.sub_pcd(argc, argv);
  cout << "finish\n";
  return 0;
}
